tests/AggregationBlogPost/article.md

```diff
@@ -0,0 +1,713 @@
+extracting complex expressions or stages into separate methods, don't keep those
+internal, but make them public as shown above. For one, this allows reusing them
+in other places, but more importantly, it allows testing them in isolation.
+
+Think about the complete pipeline shown at the beginning: you would most likely
+have tests to ensure you're receiving the correct result, but if that test fails
+you'll be forced to evaluate the entire pipeline to figure out what's wrong.
+This would include all the logic included in the pipeline. By extracting complex
+parts into separate methods, you can test those in isolation under controlled
+conditions, knowing that they behave as you intended. Looking at the last
+example where we've extracted the expression that computes the duration between
+two dates, you can now easily verify that all the individual parts of that
+complex expression work as expected.
+
+Just beware of premature abstractions: extract complex logic with the goal of
+making your code more readable and testable, not to be able to reuse it across
+your entire code base.
+
+## Builder Internals
+
+In the examples you've seen so far, we've only ever looked at stages,
+expressions, and of course the `Pipeline` objects. But how do we end up with an
+aggregation pipeline that the server understands? The objects created with the
+various factory methods are value holder objects. They don't contain any logic,
+but they allow us to know what types an expression will resolve to.
+
+For an example, let's take the `$hour` operator. When you use
+`Expression::hour`, you will receive an instance of an `HourOperator` class.
+This class implements an `OperatorInterface`, telling the builder that this is
+an operator that can be used in aggregation pipeline. It also implements a
+`ResolvesToInt` interface, as we always know that evaluating the expression
+results in an integer value. The required `date` parameter of the operator is a
+date, which in the builder is one of many things. It could be a BSON
+`UTCDateTime` instance, but it could also be the result of any operator that
+returns a date, e.g. `$dateFromString`.
+
+Now that we know about these value holder objects, we still need to make sure
+the server knows what we're talking about. When you call `Collection::aggregate`
+with the pipeline you built, what happens internally to it? Here, a series of
+encoders springs into action. We use a single entry point, the `BuilderEncoder`
+class. This class contains multiple encoders that are able to handle all
+pipeline stages, operators, and accumulators and transform them into their BSON
+representations.
+
+This allows us to keep the logic customizable. For example, the Doctrine MongoDB
+ODM allows users to specify different names for fields in the database. In turn,
+it needs to convert the name of a property in the mapped class to the name of
+the field in the database. Such a feature could easily be built by creating a
```

Comment:
I'll defer to you if "easily" makes sense here. I imagine it wouldn't be for someone that isn't an ODM contributor.

Reply:
Excellent point. I used this opportunity and removed the other occurrences of "easily" as well.